rdr_service/tools/tool_libs/unconsent.py

```diff
@@ -0,0 +1,119 @@
+import argparse
+
+from sqlalchemy import and_
+from sqlalchemy.orm import Session
+
+from rdr_service.model.code import Code
+from rdr_service.model.consent_file import ConsentFile
+from rdr_service.model.participant_summary import ParticipantSummary
+from rdr_service.model.questionnaire import QuestionnaireQuestion
+from rdr_service.model.questionnaire_response import QuestionnaireResponse, QuestionnaireResponseAnswer,\
+    QuestionnaireResponseExtension
+from rdr_service.model.utils import from_client_participant_id
+from rdr_service.tools.tool_libs.tool_base import cli_run, logger, ToolBase
+
+
+tool_cmd = 'unconsent'
+tool_desc = 'Remove participants that we have received ConsentPII payloads for, but have not actually consented'
+
+
+class UnconsentTool(ToolBase):
+
+    def run(self):
+        super(UnconsentTool, self).run()
+
+        with self.get_session() as session:
+            for participant_id in self._load_participant_ids():
+                # Retrieve the participant summary
+                summary_query = session.query(ParticipantSummary).filter(
+                    ParticipantSummary.participantId == participant_id
+                )
+                if not self.args.dry_run:
+                    # Obtain a lock on the participant summary to prevent possible
+                    # race conditions with incoming responses
+                    summary_query = summary_query.with_for_update()
```

Comment:
This may be a non-issue or may expose another race condition if it's possible for a consent to arrive during the unconsent check, but is it preferable to confirm if the participant has a signed consent first and needs the crud operation (continue to next pid if not) before obtaining a lock?

Reply:
We'd need the lock before the check. If we do the check and then lock, then we'd potentially have received (or be in the middle of receiving) a consent after the check but before the lock was obtained. This way we get the lock first and there isn't any chance for anything to be saved between the check and the removal of the data.

This strategy locks participants that wouldn't need to be locked, but the check is quick so wouldn't hold the lock for very long.

Reply:
After some offline discussion updated to lock Participant record instead, matching what the QuestionnaireResponseDao does. Thanks for helping find that 😄